myia/prim/shape_inferrers.py

```diff
@@ -108,27 +108,54 @@ def provably_equivalent(self, other):
         return type(self) == type(other)
 
 
-def find_matching_shape(shps):
-    """Returns a shape that matches all shapes in `shps`."""
-    shp = shps[0]
-    shps = shps[1:]
+@overload
+def _generalize_shape(s1: ListShape, s2):
+    return ListShape(_generalize_shape(s1.shape, s2.shape))
```

Comment:
All of these don't check that s2 is the right type.

Reply:
It's tested in `find_matching_shape` but it won't work recursively, that's true. Fixed.